File: src/ts/sync/sync-message.ts
import { v4 as uuid } from 'uuid';
import { ISyncMessage } from './i-sync-message';

export class SyncMessage<T> implements ISyncMessage<T> {
    readonly id: string;

    constructor(
        public data: T,
        public message: string = ''
    ) {
        this.id = uuid();
    }
}

File: src/dotnet/Sync/SyncMessage.cs
namespace Sync;
public class SyncMessage<T> : ISyncMessage<T>
{
    public Guid Id { get; private set; }
    public T Data { get; set; }
    public ActionType Action { get; set; }
    public string Message { get; set; }

    public SyncMessage(
        T data,
        ActionType action,
        string? message = null
    ) {
        Id = Guid.NewGuid();
        Action = action;
        Data = data;
        Message = message ?? string.Empty;
    }
}